datastore/storagebackend/postgresql/gettsattrgroups.go

```diff
@@ -0,0 +1,326 @@
+// instances that match a unique combination of database values corresponding to cols.
+// All attributes, including those in cols, are set to the actual values found in the database.
+// Returns nil upon success, otherwise error.
+func getTSAttrGroupsIncInstances(
+	db *sql.DB, cols []string, groups *[]*datastore.TSMdataGroup) error {
+	allCols := getTSAllPBNames() // get all protobuf names of TSMetadata message
+
+	goNames, err := getTSGoNamesFromPBNames(cols)
+	if err != nil {
+		return fmt.Errorf("getTSGoNamesFromPBNames() failed: %v", err)
+	}
+
+	// query database for all columns in time_series, ordered by cols
+	allColsS := strings.Join(allCols, ",")
+	colsS := strings.Join(cols, ",")
+	query := fmt.Sprintf("SELECT %s FROM time_series ORDER BY %s", allColsS, colsS)
+	rows, err := db.Query(query)
+	if err != nil {
+		return fmt.Errorf("db.Query() failed: %v", err)
+	}
+	defer rows.Close()
+
+	// aggregate rows into groups
+	currInstances := []*datastore.TSMetadata{} // initial current instance set
+	for rows.Next() {
+		// extract tsMdata from current result row
+		tsMdata, err := scanTsMdata(rows, allCols)
+		if err != nil {
+			return fmt.Errorf("scanTsMdata() failed: %v", err)
+		}
+
+		if len(currInstances) > 0 { // check if we should create a new current instance set
+			equal, err := tsMdataEqual(tsMdata, currInstances[0], goNames)
+			if err != nil {
+				return fmt.Errorf("tsMdataEqual() failed: %v", err)
+			}
+
+			if !equal { // ts metadata changed wrt. cols
+				// add next group with current instance set
+				currCombo, err := getCombo(currInstances[0], goNames)
+				if err != nil {
+					return fmt.Errorf("getCombo() failed (1): %v", err)
+				}
+				*groups = append(*groups, &datastore.TSMdataGroup{
+					Combo:     currCombo,
+					Instances: currInstances,
+				})
+				currInstances = []*datastore.TSMetadata{} // create a new current instance set
+			}
+		}
+
+		currInstances = append(currInstances, tsMdata) // add tsMdata to current instance set
+	}
+
+	// assert(len(currInstances) > 0)
+	// add final group with current instance set
+	currCombo, err := getCombo(currInstances[0], goNames)
+	if err != nil {
+		return fmt.Errorf("getCombo() failed (2): %v", err)
+	}
+	*groups = append(*groups, &datastore.TSMdataGroup{
+		Combo:     currCombo,
+		Instances: currInstances,
+	})
+
+	return nil
+}
+
+// getTSAttrGroupsComboOnly populates groups from cols such that each group contains a single,
+// unique combination of database values corresponding to cols. Other attributes than those in cols
+// have the default value for the type (i.e. "" for string, etc.).
+// Returns nil upon success, otherwise error.
+func getTSAttrGroupsComboOnly(db *sql.DB, cols []string, groups *[]*datastore.TSMdataGroup) error {
+	// query database for unique combinations of cols in time_series, ordered by cols
+	colsS := strings.Join(cols, ",")
+	query := fmt.Sprintf("SELECT DISTINCT %s FROM time_series ORDER BY %s", colsS, colsS)
```

Comment:
Might be good to add a comment that `cols` has been sanitized by `getTSDBColumns`, as this screamed SQL injection to me at first...

Reply:
I agree!